tutorials/nservicebus-101/lesson-5/tutorial.md

```diff
@@ -0,0 +1,184 @@
+
+Sometimes, a new release will contain a bug in handler logic that isn't found until the code is deployed. When this happens, many errors can flood into the error queue at once. At these times, it's incredibly valuable to be able to rollback to the old version of the endpoint, and then replay the messages through proven code. Then you can take the time to properly troubleshoot and fix the issue before attempting a new deployment.
+
+
+## Exercise
+
+This is an exploratory exercise, where we'll be playing with different retry options. We'll use the completed solution from the previous lesson.
+
+
+### Throw an exception
+
+For the purposes of this exercise, we'll create a specific bug in the Sales endpoint and watch what happens when we run the endpoint.
+
+1. In the **Sales** endpoint, locate the **PlaceOrderHandler**.
+2. After logging receipt of the message, throw an exception:
+
+snippet:Throw
+
+Now, run the solution.
+
+ 1. In Visual Studio's **Debug** menu, select **Detach All** so that the system keeps running, but does not break into the debugger when we throw our exception.
+ 1. In the **ClientUI** window, place an order by pressing `P`.
+
+When we do these steps, we'll see a wall of exception messages in white text, which is log level INFO, followed by one in yellow text, which is log level WARN. The exception traces in white are the failures during immediate retries, and the last trace in yellow is the failure that hands the message over to delayed retries.
+
+```no-highlight
+INFO  Sales.PlaceOrderHandler Received PlaceOrder, OrderId = e927667c-b949-47ee-8ea2-f29523909784
+WARN  NServiceBus.RecoverabilityExecutor Delayed Retry will reschedule message '53ac6836-48ef-49dd-aabb-a67c0104a2a5' after a delay of 00:00:10 because of an exception:
+System.Exception: BOOM
+   at <stack trace>
+```
+
+10 seconds later, the retries begin again, followed by another yellow trace, sending the message back to delayed retries. 20 seconds after that, another set of traces. Finally, 30 seconds after that, the final exception trace will be shown in red, which is log level ERROR. This is where NServiceBus gives up on the message and redirects it to the error queue.
+
+```no-highlight
+INFO  Sales.PlaceOrderHandler Received PlaceOrder, OrderId = e927667c-b949-47ee-8ea2-f29523909784
+ERROR NServiceBus.RecoverabilityExecutor Moving message '53ac6836-48ef-49dd-aabb-a67c0104a2a5' to the error queue 'error' because processing failed due to an exception:
+System.Exception: BOOM
+   at < stack trace>
+```
+
+
+### Modify immediate retries
+
+The only configurable option you need for immediate retries is how many of them to attempt. The default value is `5`, but you may want to set it to a higher or lower number. Many developers prefer to set it to `0` so that they can [avoid the "wall of text" effect when an exception is thrown](/samples/logging/stack-trace-cleaning/) during development, and then set it to a higher number for production use.
```

Comment:
> Many developers prefer to set it to `0` so that they can [avoid the "wall of text" effect when an exception is thrown](/samples/logging/stack-trace-cleaning/) during development, and then set it to a higher number for production use.

This is misleading. it seems to equate setting "0" to the same thing that is done in the stack-trace-cleaning sample